src/snowflake/snowpark/mock/_functions.py

```diff
@@ -609,120 +607,174 @@ def _to_timestamp(
+                    # "When an INTEGER value is cast directly to TIMESTAMP_NTZ ...
+                    # However, if the INTEGER value is stored inside a VARIANT value,
+                    # for example as shown below, then the conversion is indirect,
+                    # and is affected by the local time zone, even though the final result is TIMESTAMP_NTZ:"
+                    if caller == "timestamp_ntz":
+                        # local timestamp
+                        local_now = datetime.datetime.now(LocalTimezone.LOCAL_TZ)
+                        parsed = datetime.datetime.utcfromtimestamp(
+                            data
+                        ) + datetime.timedelta(
+                            seconds=local_now.utcoffset().total_seconds()
                         )
+                        return parsed
                     else:
-                        raise
-                parsed = datetime_data + datetime.timedelta(hours=hour_delta)
-
+                        parsed = datetime.datetime.utcfromtimestamp(data)
+                    # utc timestamps should be in utc timezone
+                    if add_timezone:
+                        parsed = parsed.replace(tzinfo=pytz.utc)
+                elif isinstance(data, str):
+                    # A string containing an integer number of seconds or milliseconds.
+                    if data.isdigit():
+                        parsed = datetime.datetime.utcfromtimestamp(
+                            convert_numeric_string_value_to_float_seconds(data)
+                        )
+                        # utc timestamps should be in utc timezone
+                        if add_timezone:
+                            parsed = parsed.replace(tzinfo=pytz.utc)
+                    # A string from which to extract a timestamp.
+                    else:
+                        parsed = dateutil.parser.parse(data)
+                # A timestamp.
+                elif isinstance(data, datetime.datetime):
+                    parsed = data
+                else:
+                    raise TypeError(
+                        f"[Local Testing] Unsupported conversion to_timestamp* of value {data} of VariantType"
+                    )
+            else:
+                raise TypeError(
+                    f"[Local Testing] Unsupported conversion to_timestamp* of data type {type(column.sf_type.datatype).__name__}"
+                )
             # Add the local timezone if tzinfo is missing and a tz is desired
             if parsed and add_timezone and parsed.tzinfo is None:
                 parsed = LocalTimezone.replace_tz(parsed)
-
-            res.append(parsed)
+            parsed = parsed + datetime.timedelta(hours=hour_delta)
+            return parsed
         except BaseException:
             if try_cast:
-                res.append(None)
+                return None
             else:
                 raise
-    return res
+
+    res = (
+        column.astype(object)
```

Comment:
Question: why do we need this astype?

Reply:
interestingly we don't need this astype, I pasted the piece of code from some stackoverflow thread.
removed it